docs/changehistory/NextVersion.md

```diff
@@ -1155,3 +1155,115 @@ were removed from the `@itwin/core-backend` package and moved to a new package,
+  db.query("SELECT IsPlanProjection, JsonProperties FROM bis.SpatialModel WHERE ECInstanceId IN (" + ids + ")");
+```
+
+The list of comma-separated Ids could be extremely long - in some cases, it might be so long that it would need to be split up into multiple queries!
+
+Now, you can bind a set of Ids as a parameter for the `IN` clause. The Ids will be serialized in a compact string format.
+
+```ts
+  const params = new QueryBinder().bindIdSet("modelIds", modelSelector.models);
+  db.query("SELECT IsPlanProjection, JsonProperties FROM bis.SpatialModel WHERE ECInstanceId IN :modelIds", params);
+```
+
+### Upgrading existing code to use the new `query` methods:
+
+The signature of the method has changed to:
+
+```ts
+query(ecsql: string, params?: QueryBinder, rowFormat?: QueryRowFormat, config?: QueryOptions): AsyncIterableIterator<any>;
+```
+
+The `rowFormat` parameter defaults to `QueryRowFormat.Array`. That format is more efficient so its use is preferred, but it differs from the previous row format. You can upgrade existing code to use the old format with minimal changes. For example, if your existing code passes query parameters as an array, change it as follows:
+
+```ts
+  // Replace this:
+  db.query("SELECT * FROM bis.Element WHERE ECInstanceId=?", ["0x1"]);
+  // With this:
+  db.query("SELECT * FROM bis.Element WHERE ECInstanceId=?", QueryBinder.from(["0x1"]), QueryRowFormat.UseJsPropertyNames);
+  // The code that accesses the properties of each row can remain unchanged.
+```
+
+Similarly, if your existing code passes an object instead of an array as the query parameter, change it as follows:
+```ts
+  // Replace this:
+  db.query("SELECT * FROM bis.Element WHERE ECInstanceId = :id", {id: "0x1"});
+  // With this:
+  db.query("SELECT * FROM bis.Element WHERE ECInstanceId=?", QueryBinder.from({id: "0x1"}), QueryRowFormat.UseJsPropertyNames);
+  // The code that accesses the properties of each row can remain unchanged.
+```
+
+### Upgrading existing code to use the new `restartQuery` methods:
+
+The parameters have changed in the same way as `query`, so they can be changed as described for `query` above.
+
+### Upgrading existing code to use the new `queryRowCount` methods:
+
+The behavior of this method has not changed, but the parameters must be provided as a [QueryBinder]($common) object instead of an array or object. Upgrade your existing code as described for `query` above.
+
+## New `ECSqlReader` API (beta)
+
+[ECSqlReader]($common) is a flexible API that is used internally by methods like [IModelDb.query]($backend). It is also useful when you want the result to include metadata. For example:
+
+```ts
+  const params = new QueryBinder().bindString("name", "CompositeUnitRefersToUnit");
+  const reader = conn.createQueryReader("SELECT ECInstanceId, Name FROM meta.ECClassDef WHERE Name=:name");
+
+  // The first time step is called, the metadata is populated.
+  while (await reader.step()) {
+    if (needMetaData)
```

Comment:
`needMetaData` and `setupColumns` are not defined. Should make more clear that at any time you can call getMetaData on the reader, or on any row.